File: CS/RichEditDocumentServer_ImageSourceExample/Form1.cs
using DevExpress.Office.Services;
using DevExpress.XtraRichEdit;
using DevExpress.XtraRichEdit.API.Native;
using DevExpress.XtraRichEdit.Export;
using System;
using DevExpress.XtraEditors;

namespace RichEditDocumentServer_ImageSourceExample
{
    public partial class Form1 : XtraForm
    {
        RichEditDocumentServer server;
        public Form1()
        {
            InitializeComponent();
            server = new RichEditDocumentServer();
            server.CreateNewDocument();
            server.Document.Images.Append
                (DocumentImageSource.FromUri("http://www.devexpress.com/i/pagetitles/xtrarichedit.png", null));
            embedImagesCheck.EditValue = true;
            server.ContentChanged += new EventHandler(richEditControl1_ContentChanged);
        }

        private void richEditControl1_ContentChanged(object sender, EventArgs e)
        {
            ReloadHtml();
        }

        private void ReloadHtml()
        {
            #region #GetHtmlText
            HtmlDocumentExporterOptions exportOptions = new HtmlDocumentExporterOptions();
            exportOptions.EmbedImages = embedImagesCheck.Checked;
            string sText = server.Document.GetHtmlText(server.Document.Range, new CustomUriProvider(), exportOptions);
            #endregion #GetHtmlText
            memoEdit1.Text = sText;
        }

        #region #documentloaded
        private void server_DocumentLoaded(object sender, EventArgs e)
        {
            IUriProviderService service = server.GetService<IUriProviderService>();
            if (service != null)
            {
                service.RegisterProvider(new CustomUriProvider());
            }
        }
        #endregion #documentloaded

        private void embedImagesCheck_CheckedChanged(object sender, EventArgs e)
        {
            if (embedImagesCheck.Checked)
                lblText.Text = "CustomUriProvider is idle.";
            else
                lblText.Text = "The CustomUriProvider.CreateImageUri method is called to write the original image uri.";
            ReloadHtml();
        }

        private void simpleButton1_Click(object sender, EventArgs e)
        {
            server.SaveDocument("result.docx", DocumentFormat.OpenXml);
            System.Diagnostics.Process.Start("result.docx");
        }
    }
}

File: VB/RichEditDocumentServer_ImageSourceExample/Form1.vb
Imports DevExpress.Office.Services
Imports DevExpress.XtraRichEdit
Imports DevExpress.XtraRichEdit.API.Native
Imports DevExpress.XtraRichEdit.Export
Imports System
Imports DevExpress.XtraEditors

Namespace RichEditDocumentServer_ImageSourceExample
    Partial Public Class Form1
        Inherits XtraForm

        Private server As RichEditDocumentServer
        Public Sub New()
            InitializeComponent()
            server = New RichEditDocumentServer()
            server.CreateNewDocument()
            server.Document.Images.Append(DocumentImageSource.FromUri("http://www.devexpress.com/i/pagetitles/xtrarichedit.png", Nothing))
            embedImagesCheck.EditValue = True
            AddHandler server.ContentChanged, AddressOf richEditControl1_ContentChanged
        End Sub

        Private Sub richEditControl1_ContentChanged(ByVal sender As Object, ByVal e As EventArgs)
            ReloadHtml()
        End Sub

        Private Sub ReloadHtml()
'            #Region "#GetHtmlText"
            Dim exportOptions As New HtmlDocumentExporterOptions()
            exportOptions.EmbedImages = embedImagesCheck.Checked
            Dim sText As String = server.Document.GetHtmlText(server.Document.Range, New CustomUriProvider(), exportOptions)
'            #End Region ' #GetHtmlText
            memoEdit1.Text = sText
        End Sub

        #Region "#documentloaded"
        Private Sub server_DocumentLoaded(ByVal sender As Object, ByVal e As EventArgs)
            Dim service As IUriProviderService = server.GetService(Of IUriProviderService)()
            If service IsNot Nothing Then
                service.RegisterProvider(New CustomUriProvider())
            End If
        End Sub
        #End Region ' #documentloaded

        Private Sub embedImagesCheck_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles embedImagesCheck.CheckedChanged
            If embedImagesCheck.Checked Then
                lblText.Text = "CustomUriProvider is idle."
            Else
                lblText.Text = "The CustomUriProvider.CreateImageUri method is called to write the original image uri."
            End If
            ReloadHtml()
        End Sub

        Private Sub simpleButton1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles simpleButton1.Click
            server.SaveDocument("result.docx", DocumentFormat.OpenXml)
            System.Diagnostics.Process.Start("result.docx")
        End Sub
    End Class
End Namespace
